### ISSUE TITLE
Inconsistent behavior between Normalizers when input is Dataframe

### ISSUE THREAD

**Issue by alicanb** [matplotlib/matplotlib#16400]
### Bug report

**Bug summary**
It seems matplotlib.colors.Normalize can work with pandas.Series inputs, whereas LogNorm can't.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection
from matplotlib.colors import ListedColormap, BoundaryNorm, Normalize, LogNorm
import pandas as pd

x = np.linspace(0, 3 * np.pi, 500)
y = np.sin(x)
dydx = np.linspace(0, 3 * np.pi, 500) +1  # first derivative
df= pd.DataFrame({'x':x,'y':y,'dydx':dydx})
points = np.array([df['x'], df['y']]).T.reshape(-1, 1, 2)
segments = np.concatenate([points[:-1], points[1:]], axis=1)
fig, axs = plt.subplots()

# Create a continuous norm to map from data points to colors
norm = LogNorm(df['dydx'].min(), df['dydx'].max())
lc = LineCollection(segments, cmap='viridis', norm=norm)
# Set the values used for colormapping
lc.set_array(df['dydx'])
lc.set_linewidth(2)
line = axs.add_collection(lc)
fig.colorbar(line, ax=axs)
axs.set_xlim(x.min(), x.max())
axs.set_ylim(-1.1, 1.1)
plt.show()
```

**Actual outcome**

```
TypeError: '<=' not supported between instances of 'SingleBlockManager' and 'int'
```

**Expected outcome**

Not sure about what's expected, but if I do `Normalize` instead of `LogNorm` it works, if I do `lc.set_array(df['dydx'].to_numpy())`  instead of `lc.set_array(df['dydx'])` it also works. It's the combination of LogNorm and pd.Series is the problem.

**Matplotlib version**

version 3.1.1
module://ipykernel.pylab.backend_inline

**Comment by chowington** [comment]
@alicanb Thanks for reporting this; I experienced exactly the same thing on legacy code. Hope it gets fixed.

**Comment by jklymak** [comment]
Can someone update this with something that works on modern matplotlib?  Even stripping all the pandas out this doesn't work for me...

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection
from matplotlib.colors import ListedColormap, BoundaryNorm, Normalize, LogNorm

x = np.linspace(0, 3 * np.pi, 500)
y = np.sin(x)
dydx = np.linspace(0, 3 * np.pi, 500) +1  # first derivative
points = np.array([x, y]).T.reshape(-1, 1, 2)
segments = np.concatenate([points[:-1], points[1:]], axis=1)

fig, axs = plt.subplots()

# Create a continuous norm to map from data points to colors
norm = Normalize(np.min(dydx), np.max(dydx))
lc = LineCollection(segments, cmap='viridis', norm=norm)
# Set the values used for colormapping
lc.set_array(dydx)
line = axs.add_collection(lc)
fig.colorbar(line, ax=axs)
axs.set_xlim(x.min(), x.max())
axs.set_ylim(-1.1, 1.1)
plt.show()
```

yields:

```
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/lib/matplotlib/testPandas.py", line 19, in <module>
    line = axs.add_collection(lc)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 2067, in add_collection
    datalim = collection.get_datalim(self.transData)
  File "/Users/jklymak/matplotlib/lib/matplotlib/collections.py", line 281, in get_datalim
    return mpath.get_path_collection_extents(
  File "/Users/jklymak/matplotlib/lib/matplotlib/path.py", line 1077, in get_path_collection_extents
    return Bbox.from_extents(*extents, minpos=minpos)
  File "/Users/jklymak/matplotlib/lib/matplotlib/transforms.py", line 852, in from_extents
    bbox = Bbox(np.reshape(args, (2, 2)))
  File "<__array_function__ internals>", line 5, in reshape
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 299, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 55, in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 44, in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
ValueError: cannot reshape array of size 2 into shape (2,2)
```

is `LineCollection` just broken?

**Comment by jklymak** [comment]
The full traceback is 
```
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/testPandas.py", line 22, in <module>
    fig.colorbar(line, ax=axs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/figure.py", line 1171, in colorbar
    cb = cbar.Colorbar(cax, mappable, **cb_kw)
  File "/Users/jklymak/matplotlib/lib/matplotlib/colorbar.py", line 1170, in __init__
    mappable.autoscale_None()
  File "/Users/jklymak/matplotlib/lib/matplotlib/cm.py", line 474, in autoscale_None
    self.norm.autoscale_None(self._A)
  File "/Users/jklymak/matplotlib/lib/matplotlib/colors.py", line 1558, in autoscale_None
    super().autoscale_None(np.ma.masked_less_equal(A, 0, copy=False))
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/site-packages/numpy/ma/core.py", line 2043, in masked_less_equal
    return masked_where(less_equal(x, value), x, copy=copy)
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/site-packages/numpy/ma/core.py", line 1015, in __call__
    result = self.f(da, db, *args, **kwargs)
TypeError: '<=' not supported between instances of 'SingleBlockManager' and 'int'
```

So clearly, our use of `np.ma.masked_less_equal(A, 0, copy=False))` is not liked by the data frame.  We could likely put some logic in ScalarMappable to normalize to a numpy array, but that ends up being a large game of whack-a-mole.  We also end up with issues of converting the array too early and losing the original information etc.    But in general, life is simpler if you pass us a numpy array.

**Comment by QuLogic** [comment]
This has been fixed by #20511.
